gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java

```diff
@@ -131,7 +136,19 @@
     Partition partition = new Partitioner(state).getGlobalPartition(previousWatermark);
     Histogram histogram = getHistogram(sourceEntity.getSourceEntityName(), watermarkColumn, state, partition);
 
-    String specifiedPartitions = generateSpecifiedPartitions(histogram, minTargetPartitionSize, maxPartitions,
+    // we should look if the count is too big, cut off early if count exceeds the limit, or bucket size is too large
+    Histogram histogramActual = new Histogram();
+    for (HistogramGroup group: histogram.getGroups()) {
+      if (histogramActual.getTotalRecordCount() + group.count > state.getPropAsLong(DYNAMIC_PROBING_TOTAL_RECORDS_LIMIT, DYNAMIC_PROBING_DEFAULT_TOTAL_RECORDS_LIMIT)) {
+        break;
+      }
+      histogramActual.add(group);
+      if (histogramActual.getGroups().size() >= state.getPropAsLong(DYNAMIC_PROBING_TOTAL_BUCKETS_LIMIT, DYNAMIC_PROBING_DEFAULT_TOTAL_BUCKETS_LIMIT)) {
```

Comment:
Why have the bucket limit here?

Reply:
@htran1 do we have total query limits? I think each bucket is a single query, if we have total limits for querying count, we can set this bucket number.

Reply:
The query limit is enforced by `salesforce.dynamicProbingLimit`. The loop here doesn't execute any queries, so the limit is not required here.